File: lib/rgfa/line/segment/references.rb
module RGFA::Line::Segment::References

  # References to the graph lines which involve the segment as dovetail overlap
  # @return [Array<RGFA::Line::Edge>] an array of lines; the lines themselves
  #   can be modified, but the array is frozen
  def dovetails
    dovetails_L + dovetails_R
  end

  # References to the graph lines which involve the segment as dovetail overlap
  # @param extremity [:L,:R] left of right extremity of the segment
  # @return [Array<RGFA::Line::Edge>] an array of lines; the lines themselves
  #   can be modified, but the array is frozen
  def dovetails_of_end(extremity)
    send(:"dovetails_#{extremity}")
  end

  # References to the graph lines which involve the segment as dovetail overlap
  def gaps
    gaps_L + gaps_R
  end

  # References to the graph lines which involve the segment as dovetail overlap
  # @param extremity [:L,:R] left of right extremity of the segment
  def gaps_of_end(extremity)
    send(:"gaps_#{extremity}")
  end

  # References to graph edges (C lines for GFA1, E for GFA2) which involve the
  # segment in a containment relationship.
  def containments
    edges_to_contained + edges_to_containers
  end

  # Computes the connectivity of a segment from its number of dovetail overlaps.
  #
  # @return [Array<conn_symbol,conn_symbol>]
  #  conn. symbols respectively of the :L and :R ends of +segment+.
  #
  # <b>Connectivity symbol:</b> (+conn_symbol+)
  # - Let _n_ be the number of links to an end (+:L+ or +:R+) of a segment.
  #   Then the connectivity symbol is +:M+ if <i>n > 1</i>, otherwise _n_.
  #
  def connectivity
    if !connected?
      raise RGFA::ArgumentError,
        "Cannot compute the connectivity of #{self}\n"+
        "Segment is not connected to a RGFA instance"
    end
    connectivity_symbols(dovetails_L.size, dovetails_R.size)
  end

  # List of dovetail-neighbours of a segment
  # @return [Array<RGFA::Line::Segment>] segments connected to the current
  #   segment by dovetail overlap relationships (L lines for GFA1,
  #   dovetail-representing E lines for GFA2)
  def neighbours
    dovetails.map{|l|l.other(self)}.uniq
  end

  # List of dovetail-neighbours of a segment
  # @return [Array<RGFA::Line::Segment>] segments connected to the current
  #   segment by dovetail overlap relationships (L lines for GFA1,
  #   dovetail-representing E lines for GFA2)
  # @param extremity [:L,:R] left of right extremity of the segment
  def neighbours_of_end(extremity)
    dovetails_of_end(extremity).map{|l|l.other(self)}.uniq
  end

  # List of segments which contain the segment
  # @return [Array<RGFA::Line::Segment>] segments connected to the current
  #   segment by containment relationships (C lines for GFA1,
  #   containment-representing E lines for GFA2), where the current segment is
  #   the contained segment
  def containers
    edges_to_containers.map(&:from).uniq
  end

  # List of segments which are contained in the segment
  # @return [Array<RGFA::Line::Segment>] segments connected to the current
  #   segment by containment relationships (C lines for GFA1,
  #   containment-representing E lines for GFA2), where the current segment is
  #   the container segment
  def contained
    edges_to_contained.map(&:to).uniq
  end

  # List of edges which refer to the segment
  # @return [Array<RGFA::Line::Edge>]
  def edges
    dovetails + containments + internals
  end

  # List of edges or gaps which connect to the specified segment
  # @param segment [Symbol, RGFA::Segment] a segment name or instance
  # @param collection [Symbol] <i>(defaults to: +edges+)</i> which edges or gaps
  #   shall be considered; a method of segment which returns an array of
  #   instances which respond to the method +other+ (e.g. edges, gaps,
  #   containments, edges_to_contained; dovetails_L; gaps_L; etc)
  # @return [Array<RGFA::Line::Edge, RGFA::Line::Gap>]
  def relations_to(segment, collection = :edges)
    case segment
    when Symbol
      relations_to_symbol(segment, collection)
    when RGFA::Line
      relations_to_line(segment, collection)
    end
  end

  # List of edges or gaps which connect to the specified oriented segment
  # @param orientation [:+,:-] orientation of self
  # @param oriented_segment [RGFA::OrientedLine] an oriented line
  # @param collection [Symbol] <i>(defaults to: +edges+)</i> which edges or gaps
  #   shall be considered; a method of segment which returns an array of
  #   instances which respond to the method +other_oriented_segment+
  #   (e.g. edges, gaps, containments, edges_to_contained; dovetails_L;
  #   gaps_L; etc)
  # @return [Array<RGFA::Line::Edge, RGFA::Line::Gap>]
  def oriented_relations(orientation, oriented_segment, collection = :edges)
    send(collection).select do |e|
      (e.other_oriented_segment(OL[self, orientation]) rescue nil) ==
        oriented_segment.to_oriented_line
    end
  end

  # List of edges or gaps which connect to the specified segment end
  # @param orientation [:+,:-] orientation of self
  # @param oriented_segment [RGFA::OrientedLine] an oriented segment
  # @param collection [Symbol] <i>(defaults to: +edges+)</i> which edges or gaps
  #   shall be considered; a method of segment which returns an array of
  #   instances which respond to the method +other_end+
  #   (e.g. dovetails, gaps)
  # @return [Array<RGFA::Line::Edge, RGFA::Line::Gap>]
  def end_relations(extremity, segment_end, collection = :edges)
    send(collection).select do |e|
      (e.other_end([self, extremity].to_segment_end) rescue nil) ==
        segment_end.to_segment_end
    end
  end

  private

  def relations_to_symbol(segment_symbol, collection)
    send(collection).select do |e|
      e.other(self).name == segment_symbol
    end
  end

  def relations_to_line(segment, collection)
    send(collection).select do |e|
      e.other(self).eql?(segment)
    end
  end

  def connectivity_symbols(n,m)
    [connectivity_symbol(n), connectivity_symbol(m)]
  end

  def connectivity_symbol(n)
    n > 1 ? :M : n
  end

  def backreference_keys(ref, key_in_ref)
    case ref.record_type
    when :E
      [:dovetails_L, :dovetails_R, :internals,
       :edges_to_containers, :edges_to_contained]
    when :L
      [:dovetails_L, :dovetails_R]
    when :C
      (key_in_ref == :from_segment) ? [:edges_to_contained] : [:edges_to_containers]
    when :G
      [:gaps_L, :gaps_R]
    when :F
      [:fragments]
    when :P, :O
      [:paths]
    when :U
      [:sets]
    else
      []
    end
  end

end

File: gfapy/line/segment/references.py
import gfapy

class References:

  @property
  def dovetails(self):
    """
    References to the graph lines which involve the segment as dovetail overlap.

    Returns
    -------
    gfapy.line.Edge list
      A list of lines.
      The lines themselves can be modified, but the list is immutable.
    """
    return self.dovetails_L + self.dovetails_R

  def dovetails_of_end(self, extremity):
    """
    References to the graph lines which involve the segment as dovetail overlap.

    Returns
    -------
    gfapy.line.Edge list
      A list of lines.
      The lines themselves can be modified, but the list is immutable.
    """
    return getattr(self, "dovetails_{}".format(extremity))

  @property
  def gaps(self):
    """
    References to the gap lines which involve the segment.
    """
    return self.gaps_L + self.gaps_R

  def gaps_of_end(self, extremity):
    """
    References to the gap lines which involve the segment.
    """
    return getattr(self, "gaps_{}".format(extremity))

  @property
  def containments(self):
    """
    References to graph edges (C lines for GFA1, E for GFA2) which involve the
    segment in a containment relationship.
    """
    return self.edges_to_contained + self.edges_to_containers

  def _connectivity(self):
    """
    Computes the connectivity of a segment from its number of dovetail overlaps.

    Returns
    -------
    (conn_symbol,conn_symbol) list

    conn. symbols respectively of the :L and :R ends of +segment+.

    <b>Connectivity symbol:</b> (+conn_symbol+)
    - Let _n_ be the number of links to an end (+:L+ or +:R+) of a segment.
    Then the connectivity symbol is +:M+ if <i>n > 1</i>, otherwise _n_.
    """
    if not self.is_connected():
      raise gfapy.ArgumentError(
        "Cannot compute the connectivity of {}\n".format(self)+
        "Segment is not connected to a GFA instance")
    return self._connectivity_symbols(len(self.dovetails_L),
                                      len(self.dovetails_R))

  @property
  def neighbours(self):
    """
    List of dovetail-neighbours of a segment.

    Returns
    -------
    gfapy.line.Segment list
    	Segments connected to the current segment by dovetail overlap
      relationships (L lines for GFA1, dovetail-representing E lines for GFA2)
    """
    seen = set()
    return [l.other(self) for l in self.dovetails \
              if id(l) not in seen and not seen.add(id(l))]

  @property
  def neighbours_L(self):
    """
    List of dovetail-neighbours of a segment.

    Returns
    -------
    gfapy.line.Segment list
    	Segments connected to the current segment by dovetail overlap
      relationships (L lines for GFA1, dovetail-representing E lines for GFA2)
    """
    seen = set()
    return [l.other(self) for l in self.dovetails_L \
             if id(l) not in seen and not seen.add(id(l))]

  @property
  def neighbours_R(self):
    """
    List of dovetail-neighbours of a segment.

    Returns
    -------
    gfapy.line.Segment list
    	Segments connected to the current segment by dovetail overlap
      relationships (L lines for GFA1, dovetail-representing E lines for GFA2)
    """
    seen = set()
    return [l.other(self) for l in self.dovetails_R \
             if id(l) not in seen and not seen.add(id(l))]

  def neighbours_of_end(self, extremity):
    return getattr(self, "neighbours_{}".format(extremity))

  @property
  def containers(self):
    """
    List of segments which contain the segment.

    Returns
    -------
    gfapy.line.Segment list
    	Segments connected to the current segment by containment relationships
      (C lines for GFA1, containment-representing E lines for GFA2),
      where the current segment is the contained segment.
    """
    seen = set()
    return [l.from_segment for l in self.edges_to_containers \
             if id(l) not in seen and not seen.add(id(l))]

  @property
  def contained(self):
    """
    List of segments which are contained in the segment.

    Returns
    -------
    gfapy.line.Segment list
    	Segments connected to the current segment by containment relationships
      (C lines for GFA1, containment-representing E lines for GFA2),
      where the current segment is the container segment.
    """
    seen = set()
    return [l.to_segment for l in self.edges_to_contained \
             if id(l) not in seen and not seen.add(id(l))]

  @property
  def edges(self):
    """
    List of edges which refer to the segment

    Returns
    -------
    gfapy.line.Edge list
    """
    return self.dovetails + self.containments + self.internals

  def relations_to(self, segment, collection="edges"):
    if isinstance(segment, gfapy.Line):
      return [e for e in getattr(self, collection) \
          if (e.other(self) is segment)]
    else:
      return [e for e in getattr(self, collection) \
          if (e.other(self).name == segment)]

  def oriented_relations(self, orientation, oriented_segment, collection="edges"):
    return [e for e in getattr(self, collection) if \
      (e.other_oriented_segment(gfapy.OrientedLine(self, orientation), tolerant=True) == \
        oriented_segment)]

  def end_relations(self, extremity, segment_end, collection ="edges"):
    return [e for e in getattr(self, collection) if \
      (e.other_end(gfapy.SegmentEnd(self, extremity), tolerant=True) == \
        segment_end)]

  def _connectivity_symbols(self, n, m):
    return (self._connectivity_symbol(n), self._connectivity_symbol(m))

  def _connectivity_symbol(self, n):
    return "M" if n > 1 else n

  def _backreference_keys(self, ref, key_in_ref):
    if ref.record_type == "E":
      return ["dovetails_L", "dovetails_R", "internals",
              "edges_to_containers", "edges_to_contained"]
    elif ref.record_type == "L":
      return ["dovetails_L", "dovetails_R"]
    elif ref.record_type == "C":
      return ["edges_to_contained"] if (key_in_ref == "from_segment") \
        else ["edges_to_containers"]
    elif ref.record_type == "G":
      return ["gaps_L", "gaps_R"]
    elif ref.record_type == "F":
      return ["fragments"]
    elif ref.record_type == "P" or ref.record_type == "O":
      return ["paths"]
    elif ref.record_type == "U":
      return ["sets"]
    else:
      return []
